[foreach][Adam] Minimize use of intermediates to decrease peak memory

janeyx99 opened the pull request:
Starts addressing https://github.com/pytorch/pytorch/issues/97712 by
- Minimizing intermediates usage for foreach Adam
- Document the extra memory usage 
- Add comments within the code for clarity now that we reuse intermediates
- Add tests
- Did some refactoring

Next steps involve doing this for all other foreach implementations. Note that even after this change, foreach mem usage will be higher than forloop due to the fact that we have a minimum budget of 1 intermediate (to not muddle the input values) and the intermediate will be larger. For capturable, the memory usage is higher due to moving more tensors to CUDA.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104780
* #104779
* #104669

janeyx99 committed:
[foreach][Adam] Minimize use of intermediates to decrease peak memory

[ghstack-poisoned]

albanD commented on torch/optim/adam.py:
```diff
@@ -473,42 +477,51 @@ def _multi_tensor_adam(params: List[Tensor],
+            # Re-assign for clarity as we maintain minimal intermediates: we'll have
+            # step_size = - lr / (1 - beta1 ^ t) where t = num_steps
+            # bias_correction2 = 1 - beta2 ^ t
+            step_size = bias_correction1
```
That sounds good. Just mentioning as well that you can do things like `step_size = torch._foreach_reciprocal_(bias_correction1)` or `bias_correction2_sqrt = torch._foreach_sqrt_(bias_correction2)` if you prefer.

janeyx99 replied:
Nah, I'm p sure those return nothing. I had something like that before and they returned None. 
![image](https://github.com/pytorch/pytorch/assets/31798555/e95cf25b-fa5f-45a3-bebf-b9fb6f4df30e)

albanD replied:
What? :o  All our other inplace ops return the modified arg... This is very weird...

janeyx99 replied:
"foreach support is beta" i guess? I can make another issue tracking this when we wanna make foreach legit

crcrpar replied:
just so you know IIRC the current torchgen doesn't allow inplace foreach functions to return a list of Tensors (I don't remember the exact lines off the top of my head)

janeyx99 replied:
ioh that would make sense

crcrpar replied:
I re-grepped it:
https://github.com/pytorch/pytorch/blob/d8cb80e3827226eeaafd6a007018112301f89d41/torchgen/model.py#L1411-L1428

albanD commented on torch/optim/adam.py:
```diff
@@ -473,42 +477,51 @@ def _multi_tensor_adam(params: List[Tensor],
 
-            torch._foreach_addcdiv_(device_params, device_exp_avgs, denom)
+            # Folds in (admittedly ugly) 1-elem step_size math here to avoid extra param-set-sized read+write
+            # (can't fold it into addcdiv_ below because addcdiv_ requires value is a Number, not a Tensor)
```
I don't think this is true.

janeyx99 replied:
Rewrote to be cleaner...though I still didn't need addcdiv 🤔

albanD reviewed (approved):
Small stylistic nits but sound good otherwise.

janeyx99 committed:
Update on "[foreach][Adam] Minimize use of intermediates to decrease peak memory"

Starts addressing https://github.com/pytorch/pytorch/issues/97712 by
- Minimizing intermediates usage for foreach Adam
- Document the extra memory usage 
- Add comments within the code for clarity now that we reuse intermediates
- Add tests
- Did some refactoring

Next steps involve doing this for all other foreach implementations. Note that even after this change, foreach mem usage will be higher than forloop due to the fact that we have a minimum budget of 1 intermediate (to not muddle the input values) and the intermediate will be larger. For capturable, the memory usage is higher due to moving more tensors to CUDA.




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot merge

janeyx99 committed:
Update on "[foreach][Adam] Minimize use of intermediates to decrease peak memory"

Starts addressing https://github.com/pytorch/pytorch/issues/97712 by
- Minimizing intermediates usage for foreach Adam
- Document the extra memory usage 
- Add comments within the code for clarity now that we reuse intermediates
- Add tests
- Did some refactoring

Next steps involve doing this for all other foreach implementations. Note that even after this change, foreach mem usage will be higher than forloop due to the fact that we have a minimum budget of 1 intermediate (to not muddle the input values) and the intermediate will be larger. For capturable, the memory usage is higher due to moving more tensors to CUDA.




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
